patroni/scripts/wale_restore.py

```diff
@@ -0,0 +1,227 @@
+# - WALE_BACKUP_THRESHOLD_PERCENTAGE if WAL size exceeds a certain percentage of the
+
+# this script depends on an envdir defining the S3 bucket (or SWIFT dir),and login
+# credentials per WALE Documentation.
+
+# DO NOT USE with additional restore_commands; this script writes the restore command
+
+#   latest backup size
+import logging
+import os
+import psycopg2
+import subprocess
+import sys
+import argparse
+
+
+if sys.hexversion >= 0x03000000:
+    long = int
+
+logger = logging.getLogger(__name__)
+
+
+class Restore(object):
+
+    def __init__(self, scope, role, datadir, connstring, env_dir, threshold_mb, threshold_pct, use_iam):
+        self.scope = scope
+        self.role = role
+        self.master_connection = connstring
+        self.data_dir = datadir
+        self.wal_e.dir = env_dir
+        self.wal_e.threshold_mb = threshold_mb
+        self.wal_e.threshold_pct = threshold_pct
+        if use_iam == 1:
+            self.wal_e.iam_string = ' --aws-instance-profile '
+        else:
+            self.wal_e.iam_string = ''
+
+    def setup(self):
+        pass
+
+    def replica_method(self):
+        return self.create_replica_with_pg_basebackup
+
+    def replica_fallback_method(self):
+        return None
+
+    def run(self):
+        """ creates a new replica using either pg_basebackup or WAL-E """
+        method_fn = self.replica_method()
+        ret = method_fn() if method_fn else 1
+        if ret != 0 and self.replica_fallback_method() is not None:
+            ret = (self.replica_fallback_method())()
+        return ret
+
+    def create_replica_with_pg_basebackup(self):
+        try:
+            ret = subprocess.call(['pg_basebackup', '-R', '-D', '-x',
+                                   self.data_dir, '--host=' + self.master_connection['host'],
+                                   '--port=' + str(self.master_connection['port']),
+                                   '-U', self.master_connection['user']])
+        except Exception as e:
+            logger.error('Error when fetching backup with pg_basebackup: {0}'.format(e))
+            return 1
+        return ret
+
+
+class WALERestore(Restore):
+
+    def __init__(scope, role, datadir, connstring, env_dir, threshold_mb, threshold_pct, use_iam):
+        super(WALERestore, self).__init__(scope, role, datadir, connstring, env_dir, threshold_mb, threshold_pct, use_iam)
+        # check the environment variables
+        self.init_error = False
+
+    def setup(self):
+        # check that we actually have an envdir
+        if not os.path.exists(self.wal_e.dir):
+            self.init_error = True
+        else:
+            self.wal_e.cmd = 'envdir {0} wal-e {1} '.\
+                format(self.wal_e.dir, self.wal_e.iam_string)
+
+    def replica_method(self):
+        if self.should_use_s3_to_create_replica():
+            return self.create_replica_with_s3
+        return None
+
+    def replica_fallback_method(self):
```

Comment:
We don't want to fallback to pg_basebackup when wal-e is chosen as the only replica_method.